ThunderstoreCLI/Utils/DownloadCache.cs

```diff
@@ -0,0 +1,53 @@
+namespace ThunderstoreCLI.Utils;
+
+public sealed class DownloadCache
+{
+    public string CacheDirectory { get; }
+
+    private HttpClient Client { get; } = new()
+    {
+        Timeout = TimeSpan.FromHours(1)
+    };
+
+    public DownloadCache(string cacheDirectory)
+    {
+        CacheDirectory = cacheDirectory;
+    }
+
+    // Task instead of ValueTask here because these Tasks will be await'd multiple times (ValueTask does not allow that)
+    public Task<string> GetFileOrDownload(string filename, string downloadUrl)
+    {
+        string fullPath = Path.Combine(CacheDirectory, filename);
+        if (File.Exists(fullPath))
+        {
+            return Task.FromResult(fullPath);
+        }
+
+        return DownloadFile(fullPath, downloadUrl);
+    }
+
+    private async Task<string> DownloadFile(string fullpath, string downloadUrl)
+    {
+        int tryCount = 0;
+Retry:
+        try
+        {
+            var tempPath = fullpath + ".tmp";
+            // copy into memory first to prevent canceled downloads creating files on the disk
+            await using FileStream tempStream = new(tempPath, FileMode.Create, FileAccess.Write);
+            await using var downloadStream = await Client.GetStreamAsync(downloadUrl);
+            await downloadStream.CopyToAsync(tempStream);
+
+            File.Move(tempPath, fullpath);
+        }
+        catch
+        {
+            if (++tryCount == 5)
+            {
+                throw;
+            }
+            goto Retry;
```

Comment:
Not a fan on goto statements either 😂

Reply:
I think it might actually look worse if I switched it to something else, and the label and goto are only a few lines apart. Not like I particularly like them either, but it seems to be alright here